lib/vm/src/vmcontext.rs

```diff
@@ -469,207 +498,89 @@ mod test_vmglobal_definition {
+        unsafe { &self.storage.as_i32 }
     }
 
     /// Return a mutable reference to the value as an i32.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_i32_mut(&mut self) -> &mut i32 {
-        &mut *(self.storage.as_mut().as_mut_ptr() as *mut i32)
+    pub fn as_i32_mut(&mut self) -> &mut i32 {
+        unsafe { &mut self.storage.as_i32 }
     }
 
     /// Return a reference to the value as a u32.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_u32(&self) -> &u32 {
-        &*(self.storage.as_ref().as_ptr() as *const u32)
+    pub fn as_u32(&self) -> &u32 {
+        unsafe { &self.storage.as_u32 }
     }
 
     /// Return a mutable reference to the value as an u32.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_u32_mut(&mut self) -> &mut u32 {
-        &mut *(self.storage.as_mut().as_mut_ptr() as *mut u32)
+    pub fn as_u32_mut(&mut self) -> &mut u32 {
+        unsafe { &mut self.storage.as_u32 }
     }
 
     /// Return a reference to the value as an i64.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_i64(&self) -> &i64 {
-        &*(self.storage.as_ref().as_ptr() as *const i64)
+    pub fn as_i64(&self) -> &i64 {
+        unsafe { &self.storage.as_i64 }
     }
 
     /// Return a mutable reference to the value as an i64.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_i64_mut(&mut self) -> &mut i64 {
-        &mut *(self.storage.as_mut().as_mut_ptr() as *mut i64)
+    pub fn as_i64_mut(&mut self) -> &mut i64 {
+        unsafe { &mut self.storage.as_i64 }
     }
 
     /// Return a reference to the value as an u64.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_u64(&self) -> &u64 {
-        &*(self.storage.as_ref().as_ptr() as *const u64)
+    pub fn as_u64(&self) -> &u64 {
+        unsafe { &self.storage.as_u64 }
     }
 
     /// Return a mutable reference to the value as an u64.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_u64_mut(&mut self) -> &mut u64 {
-        &mut *(self.storage.as_mut().as_mut_ptr() as *mut u64)
+    pub fn as_u64_mut(&mut self) -> &mut u64 {
+        unsafe { &mut self.storage.as_u64 }
     }
 
     /// Return a reference to the value as an f32.
-    ///
-    /// # Safety
-    ///
-    /// To be defined (TODO).
-    #[allow(clippy::cast_ptr_alignment)]
-    pub unsafe fn as_f32(&self) -> &f32 {
-        &*(self.storage.as_ref().as_ptr() as *const f32)
+    pub fn as_f32(&self) -> &f32 {
+        unsafe { &self.storage.as_f32 }
```

Comment:
Is this really safe? Is it possible that the bit representation of the `f32` could be invalid (I know Wasm NaNs are a bit specific...) if another type was inside?

Reply:
IEEE floats have no invalid bit patterns, just NaNs. The one I needed to check was `i32` and `i64`. In C89 signed integers are permitted to have one invalid representation which isn't used on 2's complement machines. I decided it must be OK because transmute is used in the standard library to implement from_ne_bytes.